Switch canonical json

carolynvs opened the pull request:
Per CNAB Spec cnabio/cnab-spec#414, we want to support numbers in our canonical json representation.

The library we are currently using, github.com/docker/go/json, does not support this. So I am migrating us to the library mentioned in the spec as being compliant. This is the same library used by cnab-go in https://github.com/cnabio/cnab-go/pull/247.

I was not able to completely remove the import of the old library because TUF uses its RawMessage struct, which is a very simple wrapper around a byte array.

If we are interested we can try to get TUF to use an interface instead of a hard-coded struct type so that we can drop the dependency on the other canonical json library.

carolynvs committed:
go mod tidy

Signed-off-by: Carolyn Van Slyck <me@carolynvanslyck.com>

carolynvs committed:
Switch canonical json libraries

Per CNAB Spec https://github.com/cnabio/cnab-spec/pull/414, we want to
support numbers in our canonical json representation.

The library we are currently using, github.com/docker/go/json, does not support this.
So I am migrating us to the library mentioned in the spec as being
compliant. This is the same library used by cnab-go.

I was not able to completely remove the import of the old library
because TUF uses its RawMessage struct, which is a very simple wrapper
around a byte array.

If we are intersted we can try to get TUF to use an interface instead
of a hard-coded struct type so that we can drop the
dependency on the other canonical json library.

Signed-off-by: Carolyn Van Slyck <me@carolynvanslyck.com>

trishankatdatadog commented:
Interesting! Let's discuss at the CNAB Security meeting this week please...

carolynvs commented:
I've added it to Wednesday's agenda.

radu-matei commented:
FYI, the expiration error is now fixed by #91.

Happy to chat about this at the next meeting, but as long as _all_ CNAB tooling has the same serialization rules, it should not affect how we sign and verify bundles, although keep in mind this is a breaking change, i.e. the same bundle _could_ get a different content digest and invalidate the signature.

carolynvs committed:
Make linter happy

Rename package to not use an underscore

Signed-off-by: Carolyn Van Slyck <me@carolynvanslyck.com>

carolynvs commented:
@trishankatdatadog @radu-matei I've rebased to pick up the fix from main and made the linter happy.

trishankatdatadog commented on pkg/canonicaljson/canonicaljson.go:
```diff
@@ -0,0 +1,29 @@
+type RawMessage = rawmessage.RawMessage
+
+// Marshal returns the canonical json encoded value of v.
+func Marshal(v interface{}) ([]byte, error) {
```
Why do we need this function if we have `MarshalToRawMessage` later which checks for errors? Do we not want to check for errors all the time?

carolynvs replied:
I just pushed a commit removing MarshalToRawMessage.

carolynvs committed:
Remove MarshalToRawMessage

This wasn't necessary like I originally thought since []byte converts
cleanly to RawMessage by the compiler with any extra code.

Signed-off-by: Carolyn Van Slyck <me@carolynvanslyck.com>
